docling_core/utils/legacy.py

```diff
@@ -342,5 +349,276 @@ def _make_spans(cell: TableCell, table_item: TableItem):
     return legacy_doc
 
 
-# def legacy_to_docling_document(legacy_doc: DsDocument) -> DoclingDocument:
-#     """Convert a legacy document to DoclingDocument."""
+def legacy_to_docling_document(legacy_doc: DsDocument) -> DoclingDocument:  # noqa: C901
+    """Convert a legacy document to DoclingDocument."""
+
+    def _transform_prov(item: BaseCell) -> Optional[ProvenanceItem]:
+        """Create a new provenance from a legacy item."""
+        prov: Optional[ProvenanceItem] = None
+        if item.prov is not None and len(item.prov) > 0:
+            prov = ProvenanceItem(
+                page_no=int(item.prov[0].page),
+                charspan=tuple(item.prov[0].span),
+                bbox=BoundingBox.from_tuple(
+                    tuple(item.prov[0].bbox), origin=CoordOrigin.BOTTOMLEFT
+                ),
+            )
+        return prov
+
+    origin = DocumentOrigin(
+        mimetype="application/pdf",
+        filename=legacy_doc.file_info.filename,
+        binary_hash=legacy_doc.file_info.document_hash,
+    )
+    doc_name = Path(origin.filename).stem
+
+    doc: DoclingDocument = DoclingDocument(name=doc_name, origin=origin)
+
+    # define pages
+    if legacy_doc.page_dimensions is not None:
+        for page_dim in legacy_doc.page_dimensions:
+            page_no = int(page_dim.page)
+            size = Size(width=page_dim.width, height=page_dim.height)
+
+            doc.add_page(page_no=page_no, size=size)
+
+    # page headers
+    if legacy_doc.page_headers is not None:
+        for text_item in legacy_doc.page_headers:
+            if text_item.text is None:
+                continue
+            prov = _transform_prov(text_item)
+            doc.add_text(
+                label=DocItemLabel.PAGE_HEADER,
+                text=text_item.text,
+                parent=doc.furniture,
+            )
+
+    # page footers
+    if legacy_doc.page_footers is not None:
+        for text_item in legacy_doc.page_footers:
+            if text_item.text is None:
+                continue
+            prov = _transform_prov(text_item)
+            doc.add_text(
+                label=DocItemLabel.PAGE_FOOTER,
+                text=text_item.text,
+                parent=doc.furniture,
+            )
+
+    # footnotes
+    if legacy_doc.footnotes is not None:
+        for text_item in legacy_doc.footnotes:
+            if text_item.text is None:
+                continue
+            prov = _transform_prov(text_item)
+            doc.add_text(
+                label=DocItemLabel.FOOTNOTE, text=text_item.text, parent=doc.furniture
+            )
+
+    # main-text content
+    if legacy_doc.main_text is not None:
+        item: Optional[Union[BaseCell, BaseText]]
+
+        # collect all captions embedded in table and figure objects
+        # to avoid repeating them
+        embedded_captions: Dict[str, int] = {}
```

Comment:
Style comment: from python 3.9 (minimum required in `docling-core`) on, we can use the native types for type hints for `dict`, `list`, `bool`, `tuple` (i.e., no need to import `Dict` from `typings`).